File: 백준/Gold/13023. ABCDE/ABCDE.cs
var g = new graph(Array.ConvertAll(Console.ReadLine().Split(), int.Parse));

Console.WriteLine(g.arrive ? 1 : 0);

class graph
{
    List<List<int>> relations;
    bool[] visited;
    public bool arrive { get; private set; }

    public graph(int[] input)
    {
        //set ralations
        relations = Enumerable.Range(0, input[0]).Select(x => new List<int>()).ToList();
        for (int i = 0; i < input[1]; i++)
        {
            var _input = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);
            relations[_input[0]].Add(_input[1]);
            relations[_input[1]].Add(_input[0]);
        }
        visited = new bool[input[0]];
        arrive = false;

        for(int i = 0; i < input[0]; i++)
        {
            if(!arrive) hasFiveLoop(i, 1);
        }
    }

    void hasFiveLoop(int v, int depth)
    {
        visited[v] = true;

        if (depth is 5 || arrive)
        {
            arrive = true;
            return;
        }

        foreach (var next in relations[v])
        {
            if (!visited[next])
                hasFiveLoop(next, depth + 1);
        }

        visited[v] = false;
    }
}

File: 백준/Gold/13023. ABCDE/ABCDE.py
put = lambda: map(int, input().split());
n, m = put()
arrive = False;
relations = [[] for _ in range(n)]
visited = [False] * n

def dfs(v, depth):
    global arrive
    visited[v] = True

    if arrive or depth == 5:
        arrive = True
        return

    for i in relations[v]:
        if not visited[i]:
            dfs(i, depth + 1)

    visited[v] = False

for i in range(m):
    v1, v2 = put()
    relations[v1].append(v2)
    relations[v2].append(v1)

for i in range(n):
    dfs(i, 1)

print(1 if arrive else 0)